### PR TITLE
Alternator store ProvisionedThroughput

### PR DESCRIPTION
**Pull request by amnonh**
When users create a table using the Alternator API, they can decide if the billing is PROVISIONED of PAY_PER_REQUEST.
If the billing is set to PROVISIONED, they need to set the ProvisionedThroughput ReadCapacityUnits (RCU) and WriteCapacityUnits (WCU).

This series adds support for getting and setting the ProvisionedThroughput. The values will be stored as table extension tags.
Following how TTL is stored within the Alternator, we will use ```system:rcu_attribute``` and ```system:wcu_attribute``` for the labels.

The series adds a test that sets ProvisionedThroughput and validates that it gets the value back. It was tested with both Alternator and AWS.

This series is part of the effort to monitor, limit, and bill Alternator operations.

New code, no need to backport.

### CODE REVIEW

**Comment by tzach** [comment]
> PROVISIONED, they need to set the ProvisionedThroughput ReadCapacityUnits (RCU) and WriteCapacityUnits (WCU).

@amnonh, it is not clear from the PR description how the new tags are used.
Does ScyllaDB enforce the provisioned capacity by throttling extra messages? or, for now, just stored in the table metadata.

**Comment by amnonh** [comment]
> @amnonh, it is not clear from the PR description how the new tags are used. Does ScyllaDB enforce the provisioned capacity by throttling extra messages? or, for now, just stored in the table metadata.

The API determines that a user MUST provide ProvisionedThroughput when setting the billing to PROVISIONED.
This is an existing functionality (e.g. already supported in Alternator)

What this series adds is that values are stored and returned to the user with describe_table.

File: alternator/executor.cc [hunk 1]
```diff
@@ -443,6 +443,8 @@ static rjson::value generate_arn_for_index(const schema& schema, std::string_vie
 static rjson::value fill_table_description(schema_ptr schema, table_status tbl_status, service::storage_proxy const& proxy)
 {
```
Comment (avikivity) [inline alternator/executor.cc:444]: > tags that start with "aws:" are for internal use.

Leave aws: tags to aws.

File: alternator/executor.cc [hunk 2]
```diff
@@ -883,7 +895,13 @@ future<executor::request_return_type> executor::list_tags_of_resource(client_sta
     return make_ready_future<executor::request_return_type>(make_jsonable(std::move(ret)));
 }
 
-static void verify_billing_mode(const rjson::value& request) {
+struct billing_mode_type {
+    bool provisioned = false;
+    std::string rcu;
+    std::string wcu;
```
Comment (avikivity) [inline alternator/executor.cc:901]: Why is this a string?

File: alternator/executor.cc [hunk 3]
```diff
@@ -959,7 +980,7 @@ static future<executor::request_return_type> create_table_on_shard0(tracing::tra
     }
     builder.with_column(bytes(executor::ATTRS_COLUMN_NAME), attrs_type(), column_kind::regular_column);
 
-    verify_billing_mode(request);
+    billing_mode_type bm = verify_billing_mode(request); // here we need billing mode
```
Comment (avikivity) [inline alternator/executor.cc:983]: I don't understand the comment.

File: alternator/executor.cc (hunk 1 above)
Comment (amnonh) [reply alternator/executor.cc:444]: The benefit of following that convention is that users will know they will not use the same tags. The alternative is to develop our own convention of tags users cannot use.

File: alternator/executor.cc (hunk 2 above)
Comment (amnonh) [reply alternator/executor.cc:901]: so we can save it in a tag; the alternative is to do the int-to-string conversion when storing the value.

**Comment by amnonh** [comment]
Can we progress with this PR? it's self-contained and follows existing API.

File: alternator/executor.cc (hunk 1 above)
Comment (nyh) [reply alternator/executor.cc:444]: Don't we already have the convention of Alternator's private tags, those **not** copied from AWS, being called "alternator: something", not aws:something? What did we call the wrote isolation tag? The tag we use to store TTL? The tag to set tablets?

Comment (amnonh) [reply alternator/executor.cc:444]: system:ttl_attribute

Comment (amnonh) [reply alternator/executor.cc:444]: I don't have a strong opinion about the tag names, there's a convention in the API we could have used, I can make it a 'system:' something or 'alternator:' something

Comment (amnonh) [reply alternator/executor.cc:448]: I've renamed the labels in V3

**Comment by amnonh** [comment]
@avikivity V3 changed the tag names to use the same ```system:``` alternator use for storing TTL.

**Comment by amnonh** [comment]
@avikivity ping

**Comment by amnonh** [comment]
@nyh can you take a look?

**Comment by nyh** [comment]
> This series adds support for getting and setting the ProvisionedThroughput. The values will be stored as table extension tags. Following how TTL is stored within the Alternator, we will use `system:rcu_attribute` and `system:wcu_attribute` for the labels.

The tag name in TTL is indeed `system:ttl_attribute`, because this tag specifies which **attribute** (i.e., column name) is used for TTL. In your case, the word "attribute" is unrelated :-) Sounds like you should call these something like `system:provisioned_rcu` or something.
 
> The series adds a test that sets ProvisionedThroughput and validates that it gets the value back. It was tested with both Alternator and AWS.
> 
> This series is part of the effort to monitor, limit, and bill Alternator operations.
> 
> New code, no need to backport.

Please update docs/alternator/compatibility.md to better match the new reality - say something about the setting of BillingMode, RCU and WCU being taken from CreateTable/UpdateTable and returned on DescribeTable but are completely ignored.

File: alternator/executor.cc [hunk 4]
```diff
@@ -459,12 +465,22 @@ static rjson::value fill_table_description(schema_ptr schema, table_status tbl_s
...
+    }
+
     rjson::add(table_description["BillingModeSummary"], "LastUpdateToPayPerRequestDateTime", rjson::value(creation_date_seconds));
     // In PAY_PER_REQUEST billing mode, provisioned capacity should return 0
     rjson::add(table_description, "ProvisionedThroughput", rjson::empty_object());
-    rjson::add(table_description["ProvisionedThroughput"], "ReadCapacityUnits", 0);
-    rjson::add(table_description["ProvisionedThroughput"], "WriteCapacityUnits", 0);
+    if (tags_ptr && tags_ptr->contains(RCU_TAG_KEY)) {
+        rjson::add(table_description["ProvisionedThroughput"], "ReadCapacityUnits", std::stoi(tags_ptr->at(RCU_TAG_KEY)));
+        rjson::add(table_description["ProvisionedThroughput"], "WriteCapacityUnits", std::stoi(tags_ptr->at(WCU_TAG_KEY)));
```
Comment (nyh) [inline alternator/executor.cc:479]: Please note that users currently can write to "system:"-prefixed tags directly (this is useful, e.g., for setting write isolation or initial_tablets) so you can get something surprising like rcu set but wcu unset. I see you always check only if rcu is set, perhaps you should check both of them, otherwise this `at` can throw.
Also, std::stoi() can also throw if the value is not an integer...

Perhaps can do something like:
```c++
int rcu = 0;
int wcu = 0;
if (tags_ptr) {
    auto rcu_tag = tags_ptr.find(RCU_TAG_KEY);
    auto wcu_tag = tags_ptr.find(WCU_TAG_KEY);
    if (rcu_tag && wcu_tag) {
          // do atoi() and ignore failure (leaving rcu and wcu as 0)
          ...
          rcu = ..
          wcu = ...
```

File: alternator/executor.cc [hunk 5]
```diff
@@ -883,7 +899,13 @@ future<executor::request_return_type> executor::list_tags_of_resource(client_sta
     return make_ready_future<executor::request_return_type>(make_jsonable(std::move(ret)));
 }
 
-static void verify_billing_mode(const rjson::value& request) {
+struct billing_mode_type {
+    bool provisioned = false;
+    int64_t rcu;
```
Comment (nyh) [inline alternator/executor.cc:904]: I don't think we need to support more than 4 billion reads or writes per second, but maybe we should be prepared for the distant future, so ok :-)

File: alternator/executor.cc [hunk 6]
```diff
@@ -895,9 +917,12 @@ static void verify_billing_mode(const rjson::value& request) {
         if (!rjson::find(request, "ProvisionedThroughput")) {
             throw api_error::validation("When BillingMode=PROVISIONED, ProvisionedThroughput must be specified.");
         }
+        const rjson::value& throughput = request["ProvisionedThroughput"];
```
Comment (nyh) [inline alternator/executor.cc:920]: Please try to avoid searching for the same attribute multiple times. We already do rjson::find above so let's store it in a variable, something like
```c++
rjson::value *provisioned_throughput = rjson::find(request, "ProvisionedThroughput");
if (!provisioned_throughput) {
  ....
}
```

File: alternator/executor.cc [hunk 7]
```diff
@@ -895,9 +917,12 @@ static void verify_billing_mode(const rjson::value& request) {
...
+        return billing_mode_type{true, throughput["ReadCapacityUnits"].GetInt(), throughput["WriteCapacityUnits"].GetInt()};
```
Comment (nyh) [inline alternator/executor.cc:921]: GetInt will throw an ugly error is it's not an integer instead of an understandable error. You can use the existing function get_int_attribute() to do this for you, although this uses "int" and not the "uint64_t" you wanted to use. But you can also open-code the same thing.

File: alternator/executor.cc [hunk 8]
```diff
@@ -1120,6 +1145,10 @@ static future<executor::request_return_type> create_table_on_shard0(tracing::tra
     if (tags && tags->IsArray()) {
         update_tags_map(*tags, tags_map, update_tags_action::add_tags);
     }
+    if (bm.provisioned) {
+        tags_map[RCU_TAG_KEY] = std::to_string(bm.rcu);
+        tags_map[WCU_TAG_KEY] = std::to_string(bm.wcu);
+    }
```
Comment (nyh) [inline alternator/executor.cc:1151]: This patch only modifies CreateTable, what about UpdateTable?
And in CreateTable it only sets these tags for the base table - what about the GSIs/LSIs?

All these different cases should have tests written for them.

File: test/alternator/util.py [hunk 9]
```diff
@@ -136,9 +136,14 @@ def unique_table_name():
...
     name = unique_table_name()
+    BillingMode = 'PAY_PER_REQUEST'
+    if 'BillingMode' in kwargs:
+        BillingMode = kwargs['BillingMode']
+        del kwargs['BillingMode']
+
     print("fixture creating new table {}".format(name))
     table = dynamodb.create_table(TableName=name,
-        BillingMode='PAY_PER_REQUEST', **kwargs)
+        BillingMode=BillingMode, **kwargs)
```
Comment (nyh) [inline test/alternator/util.py:146]: Please note that PAY_PER_REQUEST should continue to be not just the default, but the option used in 99.9% of the test - because creating a PROVISIONED table on Amazon DynamoDB has the potential to be a very costly mistake!

So I'm not very happy of making it easier to mistakenly publish a test with PROVISIONED billing mode, but I guess I'll wait and see the next patch why you needed this change.

File: test/alternator/test_provisioned_throughput.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+# Test for ProvisionedThroughput
+# ProvisionedThroughput is part of a table definition
+# The following tests make sure we can get, set and update its value
+
+import pytest
+from botocore.exceptions import ClientError
+from decimal import Decimal
+from test.alternator.util import random_string, random_bytes, new_test_table
+
+def test_create_table(dynamodb):
```
Comment (nyh) [inline test/alternator/test_provisioned_throughput.py:14]: Please consider if you really want or need a separate test file just for this one test function. The existing test_table.py already has several tests involving billing mode, and how it works in createtable, describetable, and so on.

File: test/alternator/test_provisioned_throughput.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        AttributeDefinitions=AttributeDefinitions,
+        BillingMode='PROVISIONED',
+        ProvisionedThroughput=ProvisionedThroughput) as table:
+        got = table.meta.client.describe_table(TableName=table.name)['Table']
+
+        assert got['KeySchema'] == KeySchema
+        assert sorted(got['AttributeDefinitions'], key=lambda x: x["AttributeName"]) == sorted(AttributeDefinitions, key=lambda x: x["AttributeName"])
+        print(got)
+        if 'BillingMode' in got:
+            # BillingMode can be omitted
```
Comment (nyh) [inline test/alternator/test_provisioned_throughput.py:38]: This test is supposed to verify that BillingMode was **not** omitted, it shouldn't allow that.

File: test/alternator/test_provisioned_throughput.py (hunk 10 above)
Comment (nyh) [reply test/alternator/test_provisioned_throughput.py:14]: Please add a comment describing what the raison d'etre of the test - what you wrote it for. I assume it's to verify that if BillingMode is set to provisioned, this setting and the provisioned throughput set during CreateTable is saved and returned by DescribeTable

File: test/alternator/test_provisioned_throughput.py [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    }
+    with new_test_table(dynamodb,
+                        KeySchema=KeySchema,
+        AttributeDefinitions=AttributeDefinitions,
+        BillingMode='PROVISIONED',
+        ProvisionedThroughput=ProvisionedThroughput) as table:
+        got = table.meta.client.describe_table(TableName=table.name)['Table']
+
+        assert got['KeySchema'] == KeySchema
+        assert sorted(got['AttributeDefinitions'], key=lambda x: x["AttributeName"]) == sorted(AttributeDefinitions, key=lambda x: x["AttributeName"])
```
Comment (nyh) [inline test/alternator/test_provisioned_throughput.py:35]: Nitpick:  we already have extensive testing of the fact that describe_table returns the keyschema and attributedefinitions - see test_describe_table.py. So you don't really need to test the schema in this test (and didn't even need a range key in the schema).
Testing more than your test was "supposed" to be checking sometimes sounds like it's better, but often it's worse - it just obscures what this test really contributes to the test suite.

**Review by nyh — CHANGES_REQUESTED** [review]

File: test/alternator/util.py (hunk 9 above)
Comment (amnonh) [reply test/alternator/util.py:146]: I hope the test makes it clearer. By default, nothing will be changed; instead, the hard-coded value is now optional.

File: test/alternator/test_provisioned_throughput.py (hunk 11 above)
Comment (amnonh) [reply test/alternator/test_provisioned_throughput.py:38]: The comment is not clear, I'll change both that and the code

File: test/alternator/test_provisioned_throughput.py (hunk 10 above)
Comment (amnonh) [reply test/alternator/test_provisioned_throughput.py:14]: I assumed I'll add more tests with the additional functionality, please let me know if you prefer it all in an existing  file and I'll move it.

File: test/alternator/test_provisioned_throughput.py (hunk 12 above)
Comment (amnonh) [reply test/alternator/test_provisioned_throughput.py:35]: will do

File: test/alternator/test_provisioned_throughput.py (hunk 11 above)
Comment (nyh) [reply test/alternator/test_provisioned_throughput.py:38]: The comment is not the problem, it's the line of code that needs to be deleted :-) You need to just assert that `got['BillingMode']=='provisioned'` - you shouldn't accept the case that it's missing.

File: test/alternator/test_provisioned_throughput.py (hunk 10 above)
Comment (nyh) [reply test/alternator/test_provisioned_throughput.py:14]: I don't think this test framework will be great for tests that involve the actual **performance** of throughput provisioning (e.g., that if rcu=10 you only get 10 reads per second), so I don't expect you to be able to add a lot of additional tests to this file. But I'm not strongly opposed - there is no real penalty for having more files (it's not like the C++ tests where each additional source file adds a gigabyte and a whole day of compilation ;-)). So you can leave it a separate file.

File: alternator/executor.cc [hunk 13]
```diff
@@ -459,12 +465,22 @@ static rjson::value fill_table_description(schema_ptr schema, table_status tbl_s
...
     rjson::add(table_description, "BillingModeSummary", rjson::empty_object());
-    rjson::add(table_description["BillingModeSummary"], "BillingMode", "PAY_PER_REQUEST");
+    if (tags_ptr && tags_ptr->contains(RCU_TAG_KEY)) {
+        rjson::add(table_description["BillingModeSummary"], "BillingMode", "PROVISIONED");
+    } else {
+        rjson::add(table_description["BillingModeSummary"], "BillingMode", "PAY_PER_REQUEST");
+    }
+
     rjson::add(table_description["BillingModeSummary"], "LastUpdateToPayPerRequestDateTime", rjson::value(creation_date_seconds));
     // In PAY_PER_REQUEST billing mode, provisioned capacity should return 0
```
[the remark this answers is not in the document]
Comment (amnonh) [reply alternator/executor.cc:475]: done

File: alternator/executor.cc (hunk 5 above)
Comment (amnonh) [reply alternator/executor.cc:904]: changed to int

File: alternator/executor.cc (hunk 7 above)
Comment (amnonh) [reply alternator/executor.cc:921]: Changed and throw an exception if one of the values is missing

**Comment by amnonh** [comment]
@nyh Thank you for the review, V4 addresses it all with an additional test and the documentation change.
I will continue to add more functionality on-top of this PR (like table update)

File: alternator/executor.cc [hunk 14]
```diff
@@ -98,7 +102,7 @@ std::string make_jsonable::to_json() const {
 }
 
 json::json_return_type make_streamed(rjson::value&& value) {
-    // CMH. json::json_return_type uses std::function, not noncopyable_function. 
+    // CMH. json::json_return_type uses std::function, not noncopyable_function.
```
Comment (nyh) [inline alternator/executor.cc:105]: Please don't improve unrelated whitespaces in your patch. You can make it a separate patch in the same PR if it's important to you.

File: alternator/executor.cc [hunk 15]
```diff
@@ -455,22 +461,38 @@ static rjson::value fill_table_description(schema_ptr schema, table_status tbl_s
...
+    int rcu = 0;
+    int wcu = 0;
+
+    if (tags_ptr) {
+        auto rcu_tag = tags_ptr->find(RCU_TAG_KEY);
+        auto wcu_tag = tags_ptr->find(WCU_TAG_KEY);
+        if (rcu_tag != tags_ptr->end() && wcu_tag != tags_ptr->end()) {
+            try {
+                rcu = std::stoi(rcu_tag->second);
+                wcu =  std::stoi(wcu_tag->second);
```
Comment (nyh) [inline alternator/executor.cc:475]: I think there is an extra space here?

File: alternator/executor.cc [hunk 16]
```diff
@@ -455,22 +461,38 @@ static rjson::value fill_table_description(schema_ptr schema, table_status tbl_s
...
+                rjson::add(table_description["BillingModeSummary"], "BillingMode", "PROVISIONED");
```
Comment (nyh) [inline alternator/executor.cc:476]: I think better if you move this line to an "else" below in the `if (!rcu && !wcu)` and not inside this "try". It will be easier to understand, but also the current code if the tags are both set to 0 will still reach the rjson::add here and below also the second rjson::add. I think that's a mistake - you intended just one of them to happen.

File: alternator/executor.cc (hunk 6 above)
Comment (nyh) [reply alternator/executor.cc:926]: You didn't answer this comment, it's still relevant.

File: alternator/executor.cc [hunk 17]
```diff
@@ -895,9 +923,20 @@ static void verify_billing_mode(const rjson::value& request) {
         if (!rjson::find(request, "ProvisionedThroughput")) {
             throw api_error::validation("When BillingMode=PROVISIONED, ProvisionedThroughput must be specified.");
         }
+        const rjson::value& throughput = request["ProvisionedThroughput"];
+        auto rcu = get_int_attribute(throughput, "ReadCapacityUnits");
+        auto wcu = get_int_attribute(throughput, "WriteCapacityUnits");
+        if (!rcu.has_value()) {
+            throw api_error::validation("validation error detected: Value null at 'provisionedThroughput.readCapacityUnits' failed to satisfy constraint: Member must not be null");
```
Comment (nyh) [inline alternator/executor.cc:930]: nitpick: the error message doesn't need to begin with "validation error detected", it's already in the "validation error" type and will be printed as such.
And why is this message so complicated? Why not just say that ReadCapacityUnits is missing (not "null")? Did you copy this string from DynamoDB or something?

File: alternator/executor.cc [hunk 18]
```diff
@@ -959,7 +998,7 @@ static future<executor::request_return_type> create_table_on_shard0(tracing::tra
     }
     builder.with_column(bytes(executor::ATTRS_COLUMN_NAME), attrs_type(), column_kind::regular_column);
 
-    verify_billing_mode(request);
+    billing_mode_type bm = verify_billing_mode(request);
```
Comment (nyh) [inline alternator/executor.cc:1001]: nitpick: if you call this variable billing_mode instead of bm, it's use below will be more self-explanatory.

File: alternator/executor.cc [hunk 19]
```diff
@@ -1206,10 +1249,10 @@ future<executor::request_return_type> executor::update_table(client_state& clien
     elogger.trace("Updating table {}", request);
 
     static const std::vector<sstring> unsupported = {
-        "GlobalSecondaryIndexUpdates", 
+        "GlobalSecondaryIndexUpdates",
```
Comment (nyh) [inline alternator/executor.cc:1252]: Again, please avoid unrelated whitespace changes (or leave them to a different patch).

File: test/alternator/test_provisioned_throughput.py [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+
+import pytest
+from botocore.exceptions import ClientError
+from decimal import Decimal
+from test.alternator.util import random_string, random_bytes, new_test_table
+
+# When creating a table with PROVISIONED billing mode, ProvisionedThroughput must be explicitly set,
+# and the same values should be reflected when the table is described.
+# This test ensures that when a table is created with the specified billing mode and ProvisionedThroughput values,
+# the `describeTable` operation returns the correct ProvisionedThroughput values.
```
Comment (nyh) [inline test/alternator/test_provisioned_throughput.py:17]: Nitpick: the last two lines of this comment more-or-less repeat exactly wha the first two lines said...

File: test/alternator/test_provisioned_throughput.py [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+    }
+    with new_test_table(dynamodb,
+                        KeySchema=KeySchema,
+        AttributeDefinitions=AttributeDefinitions,
+        BillingMode='PROVISIONED',
+        ProvisionedThroughput=ProvisionedThroughput) as table:
+        got = table.meta.client.describe_table(TableName=table.name)['Table']
+        if 'BillingMode' in got:
+            # PROVISIONED BillingMode can be omitted, only check if it's present
+            assert got['BillingMode'] == 'PROVISIONED'
```
Comment (nyh) [inline test/alternator/test_provisioned_throughput.py:39]: Are you suggesting here (I asked before but I don't remember the answer) that Alternator returns BillingMode=PROVISIONED but DynamoDB doesn't bother to print it because it's the default?
If that's the case, I wonder why we didn't follow DynamoDB on this - but in any case if you write this in the comment, I won't have to ask again when I don't remember :-)

File: test/alternator/test_provisioned_throughput.py [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+        assert got['ProvisionedThroughput']['ReadCapacityUnits'] == ProvisionedThroughput['ReadCapacityUnits']
+        assert got['ProvisionedThroughput']['WriteCapacityUnits'] == ProvisionedThroughput['WriteCapacityUnits']
+
+# When creating a table with PROVISIONED billing mode, ProvisionedThroughput must be explicitly set,
+# and both Read and Write capacity should be present.
+# This test ensures that when a table is created with the specified billing mode both values
```
Comment (nyh) [inline test/alternator/test_provisioned_throughput.py:45]: Here again the third line just repeats what the first two lines said, but it's worse - the third line seems incomplete :-)

File: test/alternator/test_provisioned_throughput.py [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+                    { 'AttributeName': 'c4567', 'KeyType': 'RANGE' }
+        ]
+    AttributeDefinitions=[
+                { 'AttributeName': 'p123', 'AttributeType': 'S' },
+                { 'AttributeName': 'c4567', 'AttributeType': 'S' },
+    ]
+
+    ProvisionedThroughput={
+        'WriteCapacityUnits': 5
+    }
```
Comment (nyh) [inline test/alternator/test_provisioned_throughput.py:57]: a small suggestion (not critical: you can do something like
```python
for ProvisionedThroughput in [{'WriteCapacityUnits': 1}, {'ReadCapacityUnits': 5}]:
    with pytest.raises(...):
      ....
```

P.S. I would use the number 1 when possible, hopefully it will make the test cheaper to run on AWS :-)

File: test/alternator/test_provisioned_throughput.py [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+    with pytest.raises(Exception):
```
Comment (nyh) [inline test/alternator/test_provisioned_throughput.py:58]: Exception is much too broad - any syntax error or ridiculous Python problem will cause an "exception".
Please consider something like:
```python
        with pytest.raises(ClientError, match='ValidationException.*some message substring')
```

**Review by nyh — CHANGES_REQUESTED** [review]
I think the "functional" parts of the code look good, but I left some requests inline.

File: alternator/executor.cc (hunk 14 above)
Comment (amnonh) [reply alternator/executor.cc:108]: That's my IDE insists of fixing white space, it's by mistake that I pushed it

File: alternator/executor.cc (hunk 17 above)
Comment (amnonh) [reply alternator/executor.cc:1004]: I copy it from DynamoDB

File: test/alternator/test_provisioned_throughput.py (hunk 21 above)
Comment (amnonh) [reply test/alternator/test_provisioned_throughput.py:39]: I changed the comment, but maybe it's still not clear; DynamoDB doesn't add BillingMode if it's PROVISIONED as it's the default; it is marked as optional in the API documentation. I made sure the test supports both options, and I decided to always return it, I think it's clearer.

File: test/alternator/test_provisioned_throughput.py (hunk 23 above)
Comment (amnonh) [reply test/alternator/test_provisioned_throughput.py:57]: I wanted two different numbers (making sure we placed the right number at the right place) and wanted that is not zero or one, to make sure it's not some default.

File: alternator/executor.cc (hunk 16 above)
Comment (amnonh) [reply alternator/executor.cc:479]: ok, I've refactor

File: test/alternator/test_provisioned_throughput.py (hunk 23 above)
Comment (nyh) [reply test/alternator/test_provisioned_throughput.py]: Ok.

By the way, please avoid "resolving" the comments that other people made for you - they want to read your answer before the thread disappears! The person who asked the question should resolve it if the answer is satisfactory. Thanks.

File: alternator/executor.cc [hunk 25]
```diff
@@ -458,16 +464,35 @@ static rjson::value fill_table_description(schema_ptr schema, table_status tbl_s
...
+    if (is_pay_per_request) {
+        rjson::add(table_description["BillingModeSummary"], "BillingMode", "PAY_PER_REQUEST");
+    } else {
+        rjson::add(table_description["BillingModeSummary"], "BillingMode", "PROVISIONED");
+    }
     rjson::add(table_description, "ProvisionedThroughput", rjson::empty_object());
-    rjson::add(table_description["ProvisionedThroughput"], "ReadCapacityUnits", 0);
-    rjson::add(table_description["ProvisionedThroughput"], "WriteCapacityUnits", 0);
+    rjson::add(table_description["ProvisionedThroughput"], "ReadCapacityUnits", rcu);
+    rjson::add(table_description["ProvisionedThroughput"], "WriteCapacityUnits", wcu);
```
Comment (nyh) [inline alternator/executor.cc:495]: I should have asked this earlier, but does DynamoDB actually return zero ProvisionedThroughput units when in PAY_PER_REQUEST mode? Do we have a test for this case?

Also I just realized that unless I missed something, you didn't implement or test UpdateTable for billing mode. I guess you planned to do this later, in some followup?

File: alternator/executor.cc [hunk 26]
```diff
@@ -957,21 +982,39 @@ future<executor::request_return_type> executor::list_tags_of_resource(client_sta
...
-        if (!rjson::find(request, "ProvisionedThroughput")) {
+        const rjson::value *provisioned_throughput = rjson::find(request, "ProvisionedThroughput");
+        if (!provisioned_throughput) {
             throw api_error::validation("When BillingMode=PROVISIONED, ProvisionedThroughput must be specified.");
         }
+        const rjson::value& throughput = *provisioned_throughput;
+        auto rcu = get_int_attribute(throughput, "ReadCapacityUnits");
+        auto wcu = get_int_attribute(throughput, "WriteCapacityUnits");
+        if (!rcu.has_value()) {
+            throw api_error::validation("Value null at 'provisionedThroughput.readCapacityUnits' failed to satisfy constraint: Member must not be null");
```
Comment (nyh) [inline alternator/executor.cc:1008]: In my opinion, this is a really ugly error message and I would have changed it. We don't have to be compatible with the exact wording of the DynamoDB error message (just the type is critical!), and we usually aren't. It's just nice to have some common substring in the error message (e.g., "eadCapacityUnits", without the uncapitalized first letter!) to use in the raises() statement in tests.

**Review by nyh — APPROVED** [review]
Looks good. I would have reworded the ugly error message from scratch and not copy it verbatim from DynamoDB (clearly DynamoDB has some automatic error message logic that we don't have, and they didn't write that ugly message by hand). But other than that - looks good.

File: alternator/executor.cc (hunk 25 above)
Comment (amnonh) [reply alternator/executor.cc:495]: I will do the update in a follow-up PR. Once all the building blocks are in place, it's easy.

I don't have a test for PAY_PER_REQUEST, but I'll add one. We have a file for it after all

File: alternator/executor.cc (hunk 26 above)
Comment (amnonh) [reply alternator/executor.cc:1008]: I wasn't sure how closely you wanted to be to DynamoDB. In a sense, it was easy to copy their message. It's also easy when testing, but I can create something else that would work just as well with a regex

**Comment by nyh** [comment]
I forgot about this PR, and merged it now.
